### PR TITLE
Update System.CommandLine version

### PR DESCRIPTION
**Pull request by am11**
Adapting to the (ongoing) breaking changes..

### CODE REVIEW

**Comment by am11** [comment]
@adamsitnik, fyi, I was expecting a small win in crossgen2 size, which is published as NaitveAOT app, but it has 203520 bytes size regression on osx-arm64-release after this update (HEAD vs. HEAD~1). 🙁
```sh
$ ./build.sh pack -c Release
$ stat -f%z artifacts/bin/coreclr/osx.arm64.Release/crossgen2/osx-arm64/publish/crossgen2
28680536
```

File: src/coreclr/tools/InjectResource/Program.cs
Comment (adamsitnik) [inline src/coreclr/tools/InjectResource/Program.cs:10]: nit: style consistency with other symbols ctor usage

```suggestion
CliRootCommand rootCommand = new("Inject native resources into a Portable Executable image");
```

**Review by adamsitnik — APPROVED** [review]
@am11 WOW, thanks a lot for updating all the tools to the most recent System.CommandLine version!

The code looks great, I am going to hit the approve button, but I need to mark it as "NO MERGE" because we need to update few other repos that depend on S.CL and merge all the updates at once.

BTW @am11 do you have any feedback regarding the recent changes?

How did you like the switch from having multiple `Option` ctors to one and using properties to initialize Description, custom parser and default value factory?

How do you feel about the removal of `InvocationContext` and `ParseResult` being the only input for `SetAction`?

Was removing `context.ExitCode` and returning an int from action intuitive?

**Comment by adamsitnik** [comment]
> but it has 203520 bytes size regression on osx-arm64-release after this update

That is very strange. The only thing that could cause size growth are new localizations moved from the SDK to S.CL in https://github.com/dotnet/command-line-api/pull/2041 but I would definitely not expect them to take 20 MB!

**Comment by am11** [comment]
> The only thing that could cause size growth are new localizations moved from the SDK to S.CL in [dotnet/command-line-api#2041](https://github.com/dotnet/command-line-api/pull/2041)

@jkotas stripped localization resources from runtime tools in https://github.com/dotnet/runtime/pull/72478, does moving localization from SDK to S.CL have any effect of how `SatelliteResourceLanguages` (project property) is respected?

> but I would definitely not expect them to take 20 MB!

It's +200 KB difference, not 20 MB. 😉 Not a huge deal, but it was bit of a surprise. I was hoping / expecting that it will go the other way (a few hundred KBs would be shaved off with all the great refactoring, simplification, removal of dependencies / wrappers / indirections)

> BTW @am11 do you have any feedback regarding the recent changes?

Overall, APIs name, shape and classification has improved. 👍

Some feedback:
* Help does not show value placement anymore:
    ```diff
    --- before
    +++ after

    $ crossgen2 --help
    ...

    -  --inputbubbleref <inputbubbleref>                                                                    Input bubble reference file(s) to be added to bubble (any use of this option is unsupported!)  []
    +  --inputbubbleref                                                                                                         Input bubble reference file(s) to be added to bubble (any use of this option is unsupported!)  []
      --composite                                                                                          Emit a composite R2R image comprising a number of input assemblies
    -  --compositekeyfile <compositekeyfile>                                                                KeyFile(.snk) for specifying a Public Key for the composite image created
    +  --compositekeyfile                                                                                                       KeyFile(.snk) for specifying a Public Key for the composite image created
    ```
* There is no control over the order of --help and --version option; with `main` branch `crossgen2 -h` will show version at the bottom but with PR it's at the top. Perhaps, explicit `CliOption.AddHelp()` and `CliOption.AddVersion()` would improve/fix the user-experience for this? i.e. if user calls them, then API should simply disable the default one.
* Extending help has been made complicated compared to last release. See related change in `CommandLineHelpers.cs`.
    * First we need to know RootCommand will add `--help` automagically, so we need to find that by enumerating over `.Options` (after the object construction) and overwrite the value. This is not a great UX and can be improved.
* Similar to help, the API to "create rootcommand with version with option name `--version` and alias `-v`" is now more complicated. See related change in `CommandLineHelpers.cs`. 
* `CliArgument` has `HasDescription`, but `CliOption` doesn't.. We use it to create repro/replay mechanism. Hoping to have it back in `CliOption` for consistency. (I realize that it is simply: `DefaultValueFactory is not null`)
* `GetDefaultValue` is removed from `CliOption`, and `option.DefaultValueFactory(...requires cli argument.. which is internal...)`, so I had to compromise richness of the functionality.  See `CommandLineHelpers.cs` where I've skipped it for now to avoid reflection (top obtain `option.Argument`). This compromise makes the resulting repro package (`--make-repro`) a bit bigger than before.

> How did you like the switch from having multiple `Option` ctors to one and using properties to initialize Description, custom parser and default value factory?

It's good that we don't need to allocate an array for aliases `new CliOption(new [] {"--configuration", "-c"})` became `new CliOption("--configuration", "-c")`. I had mixed feeling about moving `Description` from ctor to init prop, consistent with other props, but given its frequent usage it's a bit lengthy name to type.

> How do you feel about the removal of `InvocationContext` and `ParseResult` being the only input for `SetAction`?

Love it! We don't have any usage beyond `ParseResult` and `ExitCode`. :)

File: src/coreclr/tools/Common/CommandLineHelpers.cs
Comment (am11) [inline src/coreclr/tools/Common/CommandLineHelpers.cs:218]: @adamsitnik, this was the functionality compromise. I couldn't find a way to obtain the default value of `CliOption`.

**Comment by davidwrighton** [comment]
@adamsitnik Do you have an update on when we'll be updating System.CommandLine versions?

**Review by adamsitnik — COMMENTED** [review]
@am11 could you please update to the provided version? We are very close to merging https://github.com/dotnet/sdk/pull/29131 and I would love to merge this PR as soon as the SDK gets updated.

If you don't have the time please let me know, I can take it from here.

Thanks!

**Review by adamsitnik — COMMENTED** [review]
@am11 could you please update to the provided version? We are very close to merging https://github.com/dotnet/sdk/pull/29131 and I would love to merge this PR as soon as the SDK gets updated.

If you don't have the time please let me know, I can take it from here.

Thanks!

**Comment by am11** [comment]
I'll have time over the weekend. (Having a busy week at work. 😁)

**Comment by am11** [comment]
Updated to `2.0.0-beta4.23307.1`.

File: src/coreclr/tools/Common/CommandLineHelpers.cs
Comment (adamsitnik) [reply src/coreclr/tools/Common/CommandLineHelpers.cs:218]: @am11 thank you for your feedback! I've opened https://github.com/dotnet/command-line-api/pull/2218 to address that.

File: src/coreclr/tools/Common/CommandLineHelpers.cs
Comment (adamsitnik) [inline src/coreclr/tools/Common/CommandLineHelpers.cs:210]: This can fail at runtime, we need to use the non-generic base type:

```log
Unable to cast object of type 'System.CommandLine.Help.HelpOption' to type 'System.CommandLine.CliOption`1[System.Object]'
```

```suggestion
                foreach (CliOption option in res.CommandResult.Command.Options)
```

And later use `optionResult.GetValueOrDefault<object>()` to achieve desired effect:

```cs
OptionResult? optionResult = res.GetResult(option);
if (optionResult is null || option.Name == "make-repro-path")
{

}

optionResult.GetValueOrDefault<object>();
```

File: src/coreclr/tools/r2rtest/R2RTest.csproj
Comment (adamsitnik) [inline src/coreclr/tools/r2rtest/R2RTest.csproj:22]: thank you for replacing the hardcoded version 👍

**Review by adamsitnik — COMMENTED** [review]
@am11 big thanks for working on this! I've found two things that we should address before merging. PTAL at my comments and let me know if you have the time to fix them.

File: src/coreclr/tools/Common/CommandLineHelpers.cs
Comment (am11) [reply src/coreclr/tools/Common/CommandLineHelpers.cs]: Do we need a new release with that change? From latest build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=303826&view=results

**Comment by adamsitnik** [comment]
The failure is unrelated: https://github.com/dotnet/runtime/issues/87414

Please don't merge yet as we need to figure out the right order in https://github.com/dotnet/sdk/pull/33157

**Comment by adamsitnik** [comment]
The failure is unrelated (#42518

**Pull request merged**
